### ISSUE TITLE
Powershell Task fails to run in Microsoft-Hosted Windows Agent

### ISSUE THREAD

**Issue by lapellaniz** [microsoft/azure-pipelines-tasks#14044]
## Required Information

Entering this information will route you directly to the right team and expedite traction.

**Question, Bug, or Feature?**  
*Type*: Bug

**Enter Task Name**: PowershellV2

list here (V# not needed):  
https://github.com/Microsoft/azure-pipelines-tasks/tree/master/Tasks

## Environment
- Server - Azure Pipelines

https://bddigitalhealth.visualstudio.com/IntegratedDiabetesManagement/_build/results?buildId=44380 (BD.Zack.PR)

- Agent - Hosted

Pool: Azure Pipelines
Image: windows-latest
Agent: Azure Pipelines 4

## Issue Description

error using Powershell on windows-latest hosted agent. It work on macosx-10.15.

```
##[error]The term 'pwsh.exe' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
```

```
##[error]The term 'powershell.exe' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
```

Sample use in YAML: (I've tried powershell|pwsh|PowerShell@2 task declratations)
```yaml
- powershell: |
     $nsExclusionRules = $env:NS_EXCLUSION_RULES -split ";"
   displayName: 'Configure test run settings'      
   env:
     NS_EXCLUSION_RULES: ${{ join(';', parameters.codeCoverageNamespaceExclusionRules) }}
```

### Task logs

```
2020-12-08T21:05:13.6857104Z ##[debug]Evaluating condition for step: 'Configure test run settings'
2020-12-08T21:05:13.6860142Z ##[debug]Evaluating: SucceededNode()
2020-12-08T21:05:13.6861048Z ##[debug]Evaluating SucceededNode:
2020-12-08T21:05:13.6863826Z ##[debug]=> True
2020-12-08T21:05:13.6864888Z ##[debug]Result: True
2020-12-08T21:05:13.6866040Z ##[section]Starting: Configure test run settings
2020-12-08T21:05:13.7112878Z ==============================================================================
2020-12-08T21:05:13.7113228Z Task         : PowerShell
2020-12-08T21:05:13.7113510Z Description  : Run a PowerShell script on Linux, macOS, or Windows
2020-12-08T21:05:13.7113841Z Version      : 2.177.0
2020-12-08T21:05:13.7114070Z Author       : Microsoft Corporation
2020-12-08T21:05:13.7114407Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/powershell
2020-12-08T21:05:13.7114839Z ==============================================================================
2020-12-08T21:05:15.0035289Z ##[debug]VstsTaskSdk 0.11.0 commit 7ff27a3e0bdd6f7b06690ae5f5b63cb84d0f23f4
2020-12-08T21:05:15.2355346Z ##[debug]Entering D:\a\_tasks\PowerShell_e213ff0f-5d5c-4791-802d-52ea3e7be1f1\2.177.0\powershell.ps1.
2020-12-08T21:05:15.2596130Z ##[debug]Loading resource strings from: D:\a\_tasks\PowerShell_e213ff0f-5d5c-4791-802d-52ea3e7be1f1\2.177.0\task.json
2020-12-08T21:05:15.3166746Z ##[debug]Loaded 11 strings.
2020-12-08T21:05:15.3296677Z ##[debug]SYSTEM_CULTURE: 'en-US'
2020-12-08T21:05:15.3339644Z ##[debug]Loading resource strings from: D:\a\_tasks\PowerShell_e213ff0f-5d5c-4791-802d-52ea3e7be1f1\2.177.0\Strings\resources.resjson\en-US\resources.resjson
2020-12-08T21:05:15.3805636Z ##[debug]Loaded 11 strings.
2020-12-08T21:05:15.4204432Z ##[debug]INPUT_ERRORACTIONPREFERENCE: 'stop'
2020-12-08T21:05:15.4279688Z ##[debug]INPUT_SHOWWARNINGS: 'false'
2020-12-08T21:05:15.4429850Z ##[debug] Converted to bool: False
2020-12-08T21:05:15.4512191Z ##[debug]INPUT_FAILONSTDERR: 'false'
2020-12-08T21:05:15.4564848Z ##[debug] Converted to bool: False
2020-12-08T21:05:15.4650361Z ##[debug]INPUT_IGNORELASTEXITCODE: 'false'
2020-12-08T21:05:15.4689472Z ##[debug] Converted to bool: False
2020-12-08T21:05:15.4758132Z ##[debug]INPUT_PWSH: 'true'
2020-12-08T21:05:15.4836684Z ##[debug] Converted to bool: True
2020-12-08T21:05:15.4920417Z ##[debug]INPUT_WORKINGDIRECTORY: 'D:\a\1\s'
2020-12-08T21:05:15.5164325Z ##[debug]Asserting container path exists: 'D:\a\1\s'
2020-12-08T21:05:15.5231420Z ##[debug]INPUT_TARGETTYPE: 'inline'
2020-12-08T21:05:15.5301310Z ##[debug]INPUT_SCRIPT: '$nsExclusionRules = $env:NS_EXCLUSION_RULES -split ";"
2020-12-08T21:05:15.5330701Z ##[debug]if ( $nsExclusionRules.Count -gt 0 ) {
2020-12-08T21:05:15.5369179Z ##[debug]  $runsettingsPath="D:\a\1\s\BD.DEVOPS.TEMPLATES\pipelines\pr\TestSettings.runsettings"
2020-12-08T21:05:15.5466602Z ##[debug]  $config = [xml](gc $runsettingsPath -Raw)
2020-12-08T21:05:15.5500337Z ##[debug]  $excludeList = $config.RunSettings.DataCollectionRunSettings.DataCollectors.DataCollector.Configuration.CodeCoverage.Functions.Exclude
2020-12-08T21:05:15.5531752Z ##[debug]  echo "Old exclusion list:"
2020-12-08T21:05:15.5619859Z ##[debug]  echo $excludeList
2020-12-08T21:05:15.5672644Z ##[debug]  echo "Adding namespace exclusion rules"
2020-12-08T21:05:15.5699720Z ##[debug]  foreach($rule in $nsExclusionRules) {
2020-12-08T21:05:15.5724685Z ##[debug]    $child = $config.CreateElement("Function")
2020-12-08T21:05:15.5747974Z ##[debug]    $child.AppendChild($config.CreateTextNode($rule))
2020-12-08T21:05:15.5805169Z ##[debug]    $excludeList.AppendChild($child)
2020-12-08T21:05:15.5836190Z ##[debug]  }
2020-12-08T21:05:15.6058182Z ##[debug]  echo "New exclusion list:"
2020-12-08T21:05:15.6154333Z ##[debug]  echo $excludeList
2020-12-08T21:05:15.6947116Z ##[debug]  $config.Save($runsettingsPath)
2020-12-08T21:05:15.6966062Z ##[debug]}
2020-12-08T21:05:15.6985387Z ##[debug]else{
2020-12-08T21:05:15.7050223Z ##[debug]  echo "No updates required."
2020-12-08T21:05:15.7224042Z ##[debug]}'
2020-12-08T21:05:15.7494624Z Generating script.
2020-12-08T21:05:15.8026029Z ##[debug]AGENT_VERSION: '2.177.1'
2020-12-08T21:05:15.8263022Z ##[debug]AGENT_TEMPDIRECTORY: 'D:\a\_temp'
2020-12-08T21:05:15.8296777Z ##[debug]Asserting container path exists: 'D:\a\_temp'
2020-12-08T21:05:15.9966228Z ##[debug]Leaving D:\a\_tasks\PowerShell_e213ff0f-5d5c-4791-802d-52ea3e7be1f1\2.177.0\powershell.ps1.
2020-12-08T21:05:16.0312246Z ##[debug]Caught exception from task script.
2020-12-08T21:05:16.0369308Z ##[debug]Error record:
2020-12-08T21:05:16.1053838Z ##[debug]The term 'pwsh.exe' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
2020-12-08T21:05:16.1077112Z ##[debug]    + CategoryInfo          : ObjectNotFound: (pwsh.exe:String) [], CommandNotFoundException
2020-12-08T21:05:16.1130717Z ##[debug]    + FullyQualifiedErrorId : CommandNotFoundException
2020-12-08T21:05:16.1191746Z ##[debug] 
2020-12-08T21:05:16.1274193Z ##[debug]Script stack trace:
2020-12-08T21:05:16.1367016Z ##[debug]
2020-12-08T21:05:16.1522328Z ##[debug]Exception:
2020-12-08T21:05:16.1591765Z ##[debug]System.Management.Automation.CommandNotFoundException: The term 'pwsh.exe' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
2020-12-08T21:05:16.2176162Z ##[error]The term 'pwsh.exe' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
2020-12-08T21:05:16.2198464Z ##[debug]Processed: ##vso[task.logissue type=error]The term 'pwsh.exe' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
2020-12-08T21:05:16.2201509Z ##[debug]Processed: ##vso[task.complete result=Failed]
2020-12-08T21:05:16.2489506Z ##[section]Finishing: Configure test run settings
```

## Troubleshooting
Checkout how to troubleshoot failures and collect debug logs: https://docs.microsoft.com/en-us/vsts/build-release/actions/troubleshooting

### Error logs

```
2020-12-08T21:05:16.2176162Z ##[error]The term 'pwsh.exe' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
```

```
Starting: Initialize job
Agent name: 'Azure Pipelines 2'
Agent machine name: 'WIN-R2NCQ2M5AME'
Current agent version: '2.177.1'
Operating System
Microsoft Windows Server 2019
10.0.17763
Datacenter
Virtual Environment
Environment: windows-2019
Version: 20201202.1
Included Software: https://github.com/actions/virtual-environments/blob/win19/20201202.1/images/win/Windows2019-Readme.md
Current image version: '20201202.1'
Agent running as: 'VssAdministrator'
##[debug]Triggering repository: BD.Zack. repository type: Git
Prepare build directory.
##[debug]Delete existing build directory: 'D:\a\1'
##[debug]Creating build directory: 'D:\a\1'
##[debug]Delete existing artifacts directory: 'D:\a\1\a'
##[debug]Creating artifacts directory: 'D:\a\1\a'
##[debug]Delete existing test results directory: 'D:\a\1\TestResults'
##[debug]Creating test results directory: 'D:\a\1\TestResults'
##[debug]Creating binaries directory: 'D:\a\1\b'
##[debug]Creating source directory: 'D:\a\1\s'
##[debug]Creating repository source directory: 'D:\a\1\s\BD.Zack'
##[debug]Creating repository source directory: 'D:\a\1\s\BD.Devops.Templates'
Set build variables.
Download all required tasks.
Downloading task: UseDotNet (2.175.0)
##[debug]Task 'UseDotNet' has been downloaded into 'D:\a\_tasks\UseDotNet_b0ce7256-7898-45d3-9cb5-176b752bfea6\2.175.0'.
Downloading task: NuGetToolInstaller (1.178.0)
##[debug]Task 'NuGetToolInstaller' has been downloaded into 'D:\a\_tasks\NuGetToolInstaller_2c65196a-54fd-4a02-9be8-d9d1837b7c5d\1.178.0'.
Downloading task: Bash (3.177.2)
##[debug]Task 'Bash' has been downloaded into 'D:\a\_tasks\Bash_6c731c3c-3c68-459a-a5c9-bde6e6595b5b\3.177.2'.
Downloading task: CmdLine (2.177.3)
##[debug]Task 'CmdLine' has been downloaded into 'D:\a\_tasks\CmdLine_d9bafed4-0b18-4f58-968d-86655b4d2ce9\2.177.3'.
Downloading task: DotNetCoreCLI (2.178.2)
##[debug]Task 'DotNetCoreCLI' has been downloaded into 'D:\a\_tasks\DotNetCoreCLI_5541a522-603c-47ad-91fc-a4b1d163081b\2.178.2'.
Downloading task: NuGetCommand (2.178.1)
##[debug]Task 'NuGetCommand' has been downloaded into 'D:\a\_tasks\NuGetCommand_333b11bd-d341-40d9-afcf-b32d5ce6f23b\2.178.1'.
Downloading task: XamarinAndroid (1.175.0)
##[debug]Task 'XamarinAndroid' has been downloaded into 'D:\a\_tasks\XamarinAndroid_27edd013-36fd-43aa-96a3-7d73e1e35285\1.175.0'.
Downloading task: PowerShell (2.177.0)
##[debug]Task 'PowerShell' has been downloaded into 'D:\a\_tasks\PowerShell_e213ff0f-5d5c-4791-802d-52ea3e7be1f1\2.177.0'.
Downloading task: reportgenerator (4.8.1)
##[debug]Task 'reportgenerator' has been downloaded into 'D:\a\_tasks\reportgenerator_be803a55-9253-4895-a525-be570d86f161\4.8.1'.
Downloading task: PublishCodeCoverageResults (1.178.0)
```

**Comment by max-zaytsev** [comment]
Hi @lapellaniz , 
Does this task fails only on this agent "Azure Pipelines 4?"
Do you have other tasks in this pipeline besides this one?

**Comment by lapellaniz** [comment]
I will be closing this issue down today. I tried a few troubleshooting steps (create a new simple pipeline, comment out other tasks in the pipeline). I found that one task which runs before the powershell task seems to cause this issue but ONLY on a windows agent. It works on macosx. 

We do have other pipelines and tasks which work on Azure Pipelines 4. This one failed on other agents as well which were windows.

What I found out was that when running on windows agent, our pipeline was also trying to configure the Xamarin SDK. When I removed this, the powershell task worked as expected further down the pipeline. The build still worked correctly so plan to leave this task which selects the xamarin sdk skipped on the windows agent. Since it works on macosx we will let it run.

```sh
SYMLINK=$(XamarinSDKVersion)
MONOPREFIX=/Library/Frameworks/Mono.framework/Versions/$SYMLINK
echo "##vso[task.setvariable variable=DYLD_FALLBACK_LIBRARY_PATH;]$MONOPREFIX/lib:/lib:/usr/lib:$DYLD_LIBRARY_FALLBACK_PATH"
echo "##vso[task.setvariable variable=PKG_CONFIG_PATH;]$MONOPREFIX/lib/pkgconfig:$MONOPREFIX/share/pkgconfig:$PKG_CONFIG_PATH"
echo "##vso[task.setvariable variable=PATH;]$MONOPREFIX/bin:$PATH"
sudo $AGENT_HOMEDIRECTORY/scripts/select-xamarin-sdk.sh $(XamarinSDKVersion)
```
